File: Programmers/[64063] 호텔 방 배정/ONE/64063.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    Map<Long, Long> map = new HashMap<>();
    public long[] solution(long k, long[] room_number) {
        int n = room_number.length;
        long[] answer = new long[n];

        for (int i = 0; i < n; i++)
            answer[i] = find(room_number[i]);

        return answer;
    }

    private long find(long num) {
        if (!map.containsKey(num)) {
            map.put(num, num + 1);
            return num;
        }

        long room = find(map.get(num));
        map.put(num, room);

        return room;
    }
}

File: Programmers/[64063] 호텔 방 배정/woong-jae/64063.js
class UnionFind {
    constructor() {
        this.root = new Map();
    }
    find(u) {
        if(!this.root.has(u)) this.root.set(u, u);
        if(u === this.root.get(u)) return u;
        this.root.set(u, this.find(this.root.get(u)));
        return this.root.get(u);
    }
    merge(u, v) {
        u = this.find(u);
        v = this.find(v);
        if(u === v) return;
        this.root.set(u, v);
    }
}

function solution(k, room_number) {
    const answer = [];
    const disjointSet = new UnionFind();
    
    room_number.forEach(room => {
        let emptyRoom = disjointSet.find(room);
        answer.push(emptyRoom);
        disjointSet.merge(emptyRoom, emptyRoom + 1);
    });
    
    return answer;
}